Ipam operator not working on azure

Hallo,

We are trying to the get the ipam operator up and running on azure but this is not working.
OpenShift version 4.6.15 manged install using version 1.0 of the operator
After installing and loading in the config the pod crashes.

Logs after the crash, I can mail you the full logs and config if you like.
If you need anymore information feel free to let me know. 

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1a8d9c2]

goroutine 1328 [running]:
github.com/redhat-cop/egressip-ipam-operator/controllers/egressipam/azure.(*AzureInfra).getAzureUsedIPsByCIDR(0xc0041c1c00, 0xc000395680, 0x20, 0x0, 0x0)
	/workspace/controllers/egressipam/azure/azure.go:219 +0x662
github.com/redhat-cop/egressip-ipam-operator/controllers/egressipam/azure.(*AzureInfra).GetUsedIPsByCIDR(0xc0041c1c00, 0xc000395680, 0x0, 0x0, 0x0)
	/workspace/controllers/egressipam/azure/azure.go:168 +0x50
github.com/redhat-cop/egressip-ipam-operator/controllers/egressipam.(*EgressIPAMReconciler).loadReconcileContext.func5(0xc000395680, 0xc0006da000, 0xc00058a158)
	/workspace/controllers/egressipam/egressipam_controller.go:538 +0x4b
created by github.com/redhat-cop/egressip-ipam-operator/controllers/egressipam.(*EgressIPAMReconciler).loadReconcileContext
	/workspace/controllers/egressipam/egressipam_controller.go:537 +0x1812
```

--- a/controllers/egressipam/azure/azure.go
+++ b/controllers/egressipam/azure/azure.go
@@ -320,7 +320,7 @@ func (i *AzureInfra) removeAllAzureSecondaryIPs(rc *reconcilecontext.ReconcileCo
 				}
 			}
 			networkInterface.IPConfigurations = &ipConfigurations
-			result, err := i.networkInterface.CreateOrUpdate(rc.Context, rc.Infrastructure.Status.PlatformStatus.Azure.NetworkResourceGroupName, *networkInterface.Name, networkInterface)
+			result, err := i.networkInterface.CreateOrUpdate(rc.Context, getResourceGroupFromResourceID(*networkInterface.ID), *networkInterface.Name, networkInterface)
 			if err != nil {
 				i.log.Error(err, "unable to update", "network interface", networkInterface.Name)
 				results <- err
@@ -388,7 +388,7 @@ func (i *AzureInfra) removeUnNeededAzureAssignedIPs(rc *reconcilecontext.Reconci
 				}
 			}
 			networkInterface.IPConfigurations = &ipConfigurations
-			result, err := i.networkInterface.CreateOrUpdate(rc.Context, rc.Infrastructure.Status.PlatformStatus.Azure.NetworkResourceGroupName, *networkInterface.Name, networkInterface)
+			result, err := i.networkInterface.CreateOrUpdate(rc.Context, getResourceGroupFromResourceID(*networkInterface.ID), *networkInterface.Name, networkInterface)
 			if err != nil {
 				i.log.Error(err, "unable to update", "network interface", networkInterface.Name)
 				results <- err
@@ -464,7 +464,7 @@ func (i *AzureInfra) addNeededAzureAssignedIPs(rc *reconcilecontext.ReconcileCon
 				}
 			}
 			networkInterface.IPConfigurations = &ipConfigurations
-			result, err := i.networkInterface.CreateOrUpdate(rc.Context, rc.Infrastructure.Status.PlatformStatus.Azure.NetworkResourceGroupName, *networkInterface.Name, networkInterface)
+			result, err := i.networkInterface.CreateOrUpdate(rc.Context, getResourceGroupFromResourceID(*networkInterface.ID), *networkInterface.Name, networkInterface)
 			if err != nil {
 				i.log.Error(err, "unable to update", "network interface", networkInterface.Name)
 				results <- err

--- a/controllers/egressipam/azure/azure.go
+++ b/controllers/egressipam/azure/azure.go
@@ -305,7 +305,7 @@ func (i *AzureInfra) removeAllAzureSecondaryIPs(rc *reconcilecontext.ReconcileCo
 				if *netif.Primary {
 					//load network interface
 					var err error
-					networkInterface, err = i.networkInterface.Get(rc.Context, rc.Infrastructure.Status.PlatformStatus.Azure.NetworkResourceGroupName, getNameFromResourceID(*netif.ID), "")
+					networkInterface, err = i.networkInterface.Get(rc.Context, getResourceGroupFromResourceID(*netif.ID), getNameFromResourceID(*netif.ID), "")
 					if err != nil {
 						i.log.Error(err, "unable to get", "network interface", netif)
 						results <- err
@@ -358,7 +358,7 @@ func (i *AzureInfra) removeUnNeededAzureAssignedIPs(rc *reconcilecontext.Reconci
 				if *netif.Primary {
 					//load network interface
 					var err error
-					networkInterface, err = i.networkInterface.Get(rc.Context, rc.Infrastructure.Status.PlatformStatus.Azure.NetworkResourceGroupName, getNameFromResourceID(*netif.ID), "")
+					networkInterface, err = i.networkInterface.Get(rc.Context, getResourceGroupFromResourceID(*netif.ID), getNameFromResourceID(*netif.ID), "")
 					if err != nil {
 						i.log.Error(err, "unable to get", "network interface", netif)
 						results <- err
@@ -427,7 +427,7 @@ func (i *AzureInfra) addNeededAzureAssignedIPs(rc *reconcilecontext.ReconcileCon
 				if *netif.Primary {
 					//load network interface
 					var err error
-					networkInterface, err = i.networkInterface.Get(rc.Context, rc.Infrastructure.Status.PlatformStatus.Azure.NetworkResourceGroupName, getNameFromResourceID(*netif.ID), "")
+					networkInterface, err = i.networkInterface.Get(rc.Context, getResourceGroupFromResourceID(*netif.ID), getNameFromResourceID(*netif.ID), "")
 					if err != nil {
 						i.log.Error(err, "unable to get", "network interface", netif)
 						results <- err
@@ -526,6 +526,10 @@ func GetAzureCredentialsRequestProviderSpec() *cloudcredentialv1.AzureProviderSp
 
 func getNameFromResourceID(id string) string {
 	return id[strings.LastIndex(id, "/"):]
+}
+
+func getResourceGroupFromResourceID(id string) string {
+	return strings.Split(id, "/")[3]
 }
 
 // AzureMachineProviderSpec is the type that will be embedded in a Machine.Spec.ProviderSpec field

--- a/controllers/egressipam/azure/azure.go
+++ b/controllers/egressipam/azure/azure.go
@@ -529,7 +529,7 @@ func getNameFromResourceID(id string) string {
 }
 
 func getResourceGroupFromResourceID(id string) string {
-	return strings.Split(id, "/")[3]
+	return strings.Split(id, "/")[4]
 }
 
 // AzureMachineProviderSpec is the type that will be embedded in a Machine.Spec.ProviderSpec field

--- a/controllers/egressipam/azure/azure.go
+++ b/controllers/egressipam/azure/azure.go
@@ -216,6 +216,9 @@ func (i *AzureInfra) getAzureUsedIPsByCIDR(rc *reconcilecontext.ReconcileContext
 		//get subnets
 		for _, subnet := range *result.Subnets {
 			for _, ipConfiguration := range *subnet.IPConfigurations {
+				if ipConfiguration.PrivateIPAddress == nil {
+					continue
+				}
 				IP := net.ParseIP(*ipConfiguration.PrivateIPAddress)
 				if IP == nil {
 					i.log.Error(err, "unable to parse ", "IP", *ipConfiguration.PrivateIPAddress)
@@ -341,7 +344,7 @@ func (i *AzureInfra) removeAllAzureSecondaryIPs(rc *reconcilecontext.ReconcileCo
 	return result.ErrorOrNil()
 }
 
-func (i *AzureInfra) removeUneededAzureAssignedIPs(rc *reconcilecontext.ReconcileContext) error {
+func (i *AzureInfra) removeUnNeededAzureAssignedIPs(rc *reconcilecontext.ReconcileContext) error {
 	results := make(chan error)
 	defer close(results)
 	for node, ips := range rc.FinallyAssignedIPsByNode {
@@ -363,7 +366,7 @@ func (i *AzureInfra) removeUneededAzureAssignedIPs(rc *reconcilecontext.Reconcil
 					}
 					//exclude first IP, add secondary IPs
 					for _, ipConfiguration := range *networkInterface.IPConfigurations {
-						if !*ipConfiguration.Primary {
+						if !*ipConfiguration.Primary && ipConfiguration.PrivateIPAddress != nil {
 							azureAssignedIPs = append(azureAssignedIPs, *ipConfiguration.PrivateIPAddress)
 						}
 					}
@@ -376,7 +379,7 @@ func (i *AzureInfra) removeUneededAzureAssignedIPs(rc *reconcilecontext.Reconcil
 			for _, ipConfiguration := range *networkInterface.IPConfigurations {
 				found := false
 				for i := range toBeRemovedIPs {
-					if toBeRemovedIPs[i] == *ipConfiguration.PrivateIPAddress {
+					if ipConfiguration.PrivateIPAddress != nil && toBeRemovedIPs[i] == *ipConfiguration.PrivateIPAddress {
 						found = true
 					}
 				}
@@ -432,7 +435,7 @@ func (i *AzureInfra) addNeededAzureAssignedIPs(rc *reconcilecontext.ReconcileCon
 					}
 					//exclude first IP, add secondary IPs
 					for _, ipConfiguration := range *networkInterface.IPConfigurations {
-						if !*ipConfiguration.Primary {
+						if !*ipConfiguration.Primary && ipConfiguration.PrivateIPAddress != nil {
 							azureAssignedIPs = append(azureAssignedIPs, *ipConfiguration.PrivateIPAddress)
 						}
 					}
@@ -488,7 +491,7 @@ func (i *AzureInfra) addNeededAzureAssignedIPs(rc *reconcilecontext.ReconcileCon
 
 // assigns secondary IPs to Azure machines
 func (i *AzureInfra) reconcileAzureAssignedIPs(rc *reconcilecontext.ReconcileContext) error {
-	err := i.removeUneededAzureAssignedIPs(rc)
+	err := i.removeUnNeededAzureAssignedIPs(rc)
 	if err != nil {
 		i.log.Error(err, "unable to remove uneeded IPs")
 		return err

--- a/controllers/egressipam/ipam.go
+++ b/controllers/egressipam/ipam.go
@@ -285,7 +285,7 @@ func (r *EgressIPAMReconciler) assignIPsToNodes(rc *reconcilecontext.ReconcileCo
 			minIPsPerNode := getMinKey(nodesByNumberOfAssignedIPsByCIDR[cidr])
 			if minIPsPerNode == -1 {
 				err := errors.New("Unable to find nodes for CIDR" + cidr)
-				r.Log.Error(err, "", cidr, "nodes", nodesByNumberOfAssignedIPsByCIDR[cidr])
+				r.Log.Error(err, "", "cidr", cidr, "nodes", nodesByNumberOfAssignedIPsByCIDR[cidr])
 				return map[string][]string{}, err
 			}
 			r.Log.V(1).Info("", "minIPsPerNode: ", minIPsPerNode, "for cidr", cidr)

